### PR TITLE
Discounts

### PR DESCRIPTION
**Pull request by mejango**
# Description

Allow tier's to have a discountPercent, tunable by the collection's owner iff the tier doesn't have a flag indicating that it can't increase the discount.

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/JB721TiersHookStore.sol [hunk 1]
```diff
@@ -874,6 +895,28 @@ contract JB721TiersHookStore is IJB721TiersHookStore {
...
+    /// @param discountPercent The new discount percent being applied.
+    function recordSetDiscountPercentOf(uint256 tierId, uint256 discountPercent) external override {
+        // Make sure the discount percent is within the bound.
+        if (discountPercent > JB721Constants.MAX_DISCOUNT_PERCENT) revert DISCOUNT_PERCENT_EXCEEDS_BOUNDS();
+
+        // Get a reference to the stored tier.
+        JBStored721Tier storage storedTier = _storedTierOf[msg.sender][tierId];
+
+        // Parse the flags.
+        (,,,, bool cannotIncreaseDiscountPercent) = _unpackBools(storedTier.packedBools);
```
Comment (simplemachine92) [inline src/JB721TiersHookStore.sol:939]: looks kinda crazy but in each of these unpack instances

`(
        /* bool ignoredVar */,
        /* bool ignoredVar2 */,
        /* bool ignoredVar3 */,
        /* bool ignoredVar4 */,
        bool cannotIncreaseDiscountPercent) = _unpackBools(storedTier.packedBools);`
        
    
is more readable and safe to me

File: src/JB721TiersHookStore.sol [hunk 2]
```diff
@@ -822,16 +835,24 @@ contract JB721TiersHookStore is IJB721TiersHookStore {
...
 
             // Make sure the provided tier exists (tiers cannot have a supply of 0).
             if (storedTier.initialSupply == 0) revert INVALID_TIER();
 
+            // Get a reference to the price.
+            uint256 price = storedTier.price;
+
+            // Apply a discount if needed.
+            if (storedTier.discountPercent > 0) {
+                price -= mulDiv(price, storedTier.discountPercent, JB721Constants.MAX_DISCOUNT_PERCENT);
```
Comment (simplemachine92) [inline src/JB721TiersHookStore.sol:837]: Doesn't this give half of the intended discountPercent?

File: src/JB721TiersHookStore.sol (hunk 1 above)
Comment (mejango) [reply src/JB721TiersHookStore.sol:939]: yeaahhhhhhh.

File: src/JB721TiersHookStore.sol (hunk 2 above)
Comment (mejango) [reply src/JB721TiersHookStore.sol:837]: should be full?

let's say discountPercent is JB721Constants.MAX_DISCOUNT_PERCENT/2, then it goes `price -= price/2` which is half right? 

anyways, a test would be nice. :]

**Pull request merged**